### ISSUE TITLE
SVTPrivacy might be incorrect?

### ISSUE THREAD

**Issue by zincorca** [NVIDIA/NVFlare#343]
According to these lines, I might understand it incorrectly, it seems `L115` pick only some of the `accepted` indices from `accepted` indices via `numpy.random.choice`. Causing the `accepted` and `candidate_idx` to not cover the whole indices of `delta_w`.

Later at `L119`, `delta_w[accepted]` are set with noised value and at `L120`, `delta_w[candidate_idx]` are set with zeros as in sparse value, but some of the `delta_w` values that are not in `accepted` or `candidate_idx` get passed as full clipped but non-noised values.

Is this the intended behavior?

https://github.com/NVIDIA/NVFlare/blob/0753aba7ad97e09db474e0d56ce843e4eee8285a/nvflare/app_common/filters/svt_privacy.py#L107-120

Create new array and setting new values like this lead to different result.

```python
output_w = np.zeros(delta_w.shape)
output_w[accepted] = np.clip(delta_w[accepted] + noise, a_min=-self.gamma, a_max=self.gamma)
output_w[candidate_idx] = 0.0
delta_w = output_w
```

**Comment by YuanTingHsieh** [comment]
@zincorca thanks for this finding, it is a bug.
@wyli has submitted a PR to fix it.

**Comment by zincorca** [comment]
@YuanTingHsieh @wyli  Thanks, I was trying to have someone look at at before. Is there any implication on application that was done before by this code such as https://github.com/NVIDIA/NVFlare/tree/main/examples/brats18 with differential privacy. Since this would mean some of the values are leaked and in a way having true gradients with some sparse or noised values.

Also If I understand correctly, `nvflare` use some of the code from `clara` version `4` (or the `clara` version with `pytorch`), since those were released with non-opensource code. To some extend `clara` version `<4` was written using `tensorflow`, now I am not familiar with version `<4`. It is probably closed source or at least obscure. I am not sure if the bug was there or not (`v4` might effected, idk about `v3`), so it might worth a look at.

**Comment by zincorca** [comment]
> Also If I understand correctly, `nvflare` use some of the code from `clara` version `4` (or the `clara` version with `pytorch`), since those were released with non-opensource code. To some extend `clara` version `<4` was written using `tensorflow`, now I am not familiar with version `<4`. It is probably closed source or at least obscure. I am not sure if the bug was there or not (`v4` might effected, idk about `v3`), so it might worth a look at.

Okay, I skim through nvidia `clara` container `v3`, `v2`, `v3.1.01` . Surprisingly, some of the `v3` source are open and it looks like it has the same bug that I just describe, so `v2` and `v3.1` might also be effected.
